File: Python/Demos/functions3.py
#   Author: Jo Crandall
#   CPSC 215 Function basics 3
#   Feb 9, 2022
#   Gonzaga University
#   This program demonstrates variable scope.



## A variable declared outside any function is automatically a global variable in the file
## Every function in the file can also see and use the global variable
studentName = 'Jo'

def printName():
    print(studentName)       ## here, we are inside the function printName(), but can see the variable studentName

## Remember, nothing will happen unless we call our function
printName()


## If however we declare a variable inside a function, it is only a local variable
## The variable doesn't exist outside the function
def assignID():
    studentID = 123456
    return studentID

print(assignID())


## Furthermore, other functions cannot see or use the variable
def printID():
    studentID = assignID()
    print(studentID)

printID()


## We can make a variable which is declared inside a function global by using the global keyword
## Note that to do this we need to first declare, THEN assign a value
def assignYear():
    global studentYear
    studentYear = 'Freshman'
    #global studentYear = 'Freshman'           ## this way will not work

assignYear()
print(studentYear)   ## we can see the global variable that was created inside assignYear()

## Note that studentYear is a global VARIABLE, which means it's value can still be changed anywhere in the program
## That is, global does not mean constant, it just means globally visible
#studentYear = 'Sophomore'
#print(studentYear)

## If we want a global CONSTANT (far recommended above global variables) then we use techniques we've previously seen
## 1) declare a variable at the beginning of the program that you will use in place of the value everywhere (easy to update) OR
## 2) declare a variable in a separate file completely and import that file to further insulate the simulated constant

File: C++/Demos/functions3.cpp
/*  Author: Jo Crandall
    CPSC 121 Function Basics 3 (Example 18)
    Oct 11, 2021
    Gonzaga University
    This program demonstrates variable scope, const, and static.
*/

#include <iostream>
using namespace std;

void myFunc(void);
void myOtherFunc(void);
bool yetAnotherFunc(void);
int theLastFunc(void);

int globalInt = 0;                      /// because these two variables are defined outside main(), they are GLOBAL variables
const double GLOBAL_CONST = 3.14;

int main(){

    /// global vars can be seen within any of the functions (but beware of variables with same name that have block scope)
    
    cout << "Here's the value of globalInt: " << globalInt << endl;
    myFunc();  /// this function changes the value of globalInt
    cout << "Now here's the value of globalInt: " << globalInt << endl;

    myOtherFunc();     /// this function creates the variable localInt
    //cout << "The value of localInt is: " << localInt << endl;

    /// Global constants can also be seen inside any function, but they CANNOT be changed by the code during runtime.
    bool flag = yetAnotherFunc();
    cout << "GLOBAL_CONST is 3.14? " << flag << endl;

    /// Variables and their values are destroyed between function calls, but a STATIC variable can have its value remembered
    cout << "staticInt has: " << theLastFunc() << endl;
    cout << "staticInt has: " << theLastFunc() << endl;
    cout << "staticInt has: " << theLastFunc() << endl;
    cout << "staticInt has: " << theLastFunc() << endl;

    return 0;
}

//This function sets the value of globalInt to 100.
void myFunc(void){
    globalInt = 100;
}

//This function sets the value of localInt to 42.
void myOtherFunc(void){
    /// A variable defined inside a function is a LOCAL VARIABLE
    /// It can only be seen inside that function
    int localInt = 0;
    cout << "Here's the value of localInt: " << localInt << endl;
    localInt = 42;
    cout << "Now here's the value of localInt: " << localInt << endl;
}

bool yetAnotherFunc(void){
    if (GLOBAL_CONST == 3.14){
        return true;
    }
    else return false;
}

int theLastFunc(void){
    static int staticInt;
    staticInt++;
    return staticInt;
}